### PR TITLE
*: add charset information into field type

### PR DESCRIPTION
**Pull request by xiongjiwei**
### Check List 

Tests 

close https://github.com/pingcap/tidb/issues/27661

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by tisonkun** [comment]
@xiongjiwei what's the issue this PR related to?

File: types/field_type.go [hunk 1]
```diff
@@ -41,12 +41,7 @@ func NewFieldType(tp byte) *FieldType {
 		Flen:    UnspecifiedLength,
 		Decimal: UnspecifiedLength,
 	}
-	if tp != mysql.TypeVarchar && tp != mysql.TypeVarString && tp != mysql.TypeString {
-		ft.Collate = charset.CollationBin
-	} else {
-		ft.Collate = mysql.DefaultCollationName
-	}
-	// TODO: use DefaultCharsetForType to set charset and collate
+	ft.Charset, ft.Collate = DefaultCharsetForType(tp)
```
Comment (tisonkun) [inline types/field_type.go:44]: the collate size is equivalent, but we now change `Charset` from `nil` to `DefaultCharset` and update tests correspondingly. Can you elaborate the background and how it won't affect caller assumption or user facing behavior?

**Comment by xiongjiwei** [comment]
hi @tisonkun, there is no issue related to this pr for now. the original code set the collation but without charset information, this is not correct since a collation must have a charset, such as `utf8mb4_bin` -> `utf8mb4`. So it is wrong if the caller assumes a collation has `nil` charset and also it will not change any behaviors.

**Comment by tisonkun** [comment]
@xiongjiwei got it. I guess it can be regarded as a subtask of https://github.com/pingcap/tidb/issues/26812 . You can create a subtask issue there and link this PR to close that.

Rest LGTM.

**Review by xhebox — APPROVED** [review]

**Review by tisonkun — APPROVED** [review]
LGTM

**Review by tangenta — APPROVED** [review]

**Pull request merged**